### ISSUE TITLE
Associations of FRQs & place.com

### ISSUE THREAD

**Issue by akshat122805** [akshat122805/akshatCSA#14]
Problem 1: Arrays/Arraylists

This problem involved tackling a series of challenges related to arrays, particularly focusing on summing elements and analyzing the uniqueness of sums in a two-dimensional array context. The principles that I used to solve this problem all hinged on working with arrays, which are ubiquitous in our team's project. Here is an example:

for (int iteration = 0; iteration < maxIterations; iteration++) {
    List<Double> predictions = predict(x_train);
    // Update coefficients and intercept
    double interceptGradient = 0.0;
    List<Double> slopeGradients = new ArrayList<>(coefficients.size());

    for (int i = 0; i < slopeGradients.size(); i++) {
        slopeGradients.set(i, 0.0);
    }

    for (int i = 0; i < x_train.size(); i++) {
        double predicted = predictions.get(i);
        interceptGradient += predicted - y_train.get(i);
        for (int j = 0; j < coefficients.size() - 1; j++) {
            slopeGradients.set(j, slopeGradients.get(j) + (predicted - y_train.get(i)) * x_train.get(i));
        }
    }

    intercept -= learningRate * interceptGradient / x_train.size();
    for (int i = 0; i < coefficients.size() - 1; i++) {
        coefficients.set(i, coefficients.get(i) - learningRate * slopeGradients.get(i) / x_train.size());
    }
}

From our Group Final Project, specifically the Logistic Regression Model, the above is an excerpt of code that really showcases the capabilities of ArrayLists in Java. The model is used, in tandem with a linear regressor, to project a player's stats. The key part of the code that highlights the use of ArrayLists is in the train method, where we update the coefficients and intercept for the regression model. This part is crucial because it iteratively adjusts the model parameters based on the training data to improve the model's predictions.

In this excerpt, I used ArrayLists to manage and dynamically update the coefficients and slope gradients. Unlike fixed-size arrays, ArrayLists in Java can resize dynamically, which is particularly useful in machine learning contexts like ours where the number of features (and thus coefficients) can vary. Through the use of the get and set methods, we efficiently update the slope gradients and coefficients during each iteration of the training process.
Problem 2: Classes

This problem tasked me with developing the HiddenWord class, the key component of a word-guessing game where the player attempts to guess a hidden word. This assignment required me to encapsulate both data and logic within a cohesive class structure. As expected, proper class design was used widely throughout our team project, as it vital for a functioning application. Here is an example:

package com.nighthawk.spring_portfolio.nbapredictor.monte_carlo;

public class PlayerStats {
    private int points;
    private int rebounds;
    private int assists;

    // Getters and setters
    public int getPoints() {
        return points;
    }

    public void setPoints(int points) {
        this.points = points;
    }

    public int getRebounds() {
        return rebounds;
    }

    public void setRebounds(int rebounds) {
        this.rebounds = rebounds;
    }

    public int getAssists() {
        return assists;
    }

    public void setAssists(int assists) {
        this.assists = assists;
    }
}

The PlayerStats class is a great example of class desgin from our project, and a perfect example of a Plain Old Java Object (POJO). The context of PlayerStats is to hold individual basketball players' statistical data, such as points, rebounds, and assists. This class forms the base layer in our application's hierarchy, as it's the most granular level of data representation.

The fields for the statistical data are kept private, ensuring that the internal representation of a player's stats is hidden from other parts of the program, safeguarding the integrity of the data.For each private field in the PlayerStats class, there getter and setter methods. The getters allow other parts of the program to read the player's statistics, while the setters enable the modification of these stats under specific conditions defined in the program.

The PlayerStats class plays a critical role in the broader structure of our project. It serves as the building block for more complex classes like Player and GameTeam. In the Player class, PlayerStats is used to represent the individual stats of a player. The GameTeam class leverages uses Player class to form a complete team.
Problem 3: 2D Arrays

The problem aimed to have me develop an efficient way to represent and manipulate a sparse array, which is a two-dimensional array with most of its elements being zeros. The primary objectives were to retrieve specific element values from this sparse structure and to modify the array by removing columns, all while maintaining the integrity of the array's structure.

2D Arrays, in their traditional form, have not been used in our project yet. owever, we expect this to soon be remedied with the introduction of the neural network for moneyline NBA game predictions. The incorporation of neural networks will naturally lead us into the realm of 2D arrays, as they are essential in representing the complex, multi-dimensional data structures typical in machine learning models. 2D arrays, in this context, will be pivotal for handling weights, biases, and activations within the network layers. With one-hot-encoding, 2D arrays tend to get a little 0-heavy, so spare arrays could be a viable option for improving model performance.

However, there are somewhat untraditional 2D Arrays scattered through our backend. One of them would be the GameTeams class:

package com.nighthawk.spring_portfolio.nbapredictor.monte_carlo;

import java.util.List;

public class GameTeams {
    private List<Player> teamA;
    private List<Player> teamB;

    // Constructors
    public GameTeams() {}

    public GameTeams(List<Player> teamA, List<Player> teamB) {
        this.teamA = teamA;
        this.teamB = teamB;
    }

    // Getters and setters
    public List<Player> getTeamA() {
        return teamA;
    }

    public void setTeamA(List<Player> teamA) {
        this.teamA = teamA;
    }

    public List<Player> getTeamB() {
        return teamB;
    }

    public void setTeamB(List<Player> teamB) {
        this.teamB = teamB;
    }
}

The GameTeams class can appear as a 2D Array due to its two teams, which can be viewed as two rows. Within these "rows" are "columns" of players.
Problem 4: Methods and Control Structures

In this problem, my task centered around implementing a system capable of representing and managing diverse groups of numbers, each defined in distinct ways. This challenge required me to design a flexible and dynamic structure with the use of various methods and control structures. This involved creating methods that could accurately identify whether a number belonged to a specific group and applying control structures such as loops and conditional statements to navigate and manipulate these groups.

Here is an examples of methods and control structures being used in our team project:

double sum = 0;
int totalPlayers = 0;

List<Player> allPlayers = new ArrayList<>(teamsData.getTeamA());
allPlayers.addAll(teamsData.getTeamB());

for (Player player : allPlayers) {
    sum += calculateFantasyPoints(player);
    totalPlayers++;
}

double mean = sum / totalPlayers;
double squaredDifferenceSum = 0;

for (Player player : allPlayers) {
    double difference = calculateFantasyPoints(player) - mean;
    squaredDifferenceSum += difference * difference;
}

double variance = squaredDifferenceSum / totalPlayers;
return Math.sqrt(variance);

The above excerpt is from the MonteCarloSimulator class in our NBA Predictor project, which is designed for simulating and predicting fantasy points in NBA games. The calculateStandardDeviation is crucial for accurately capturing the variability in player performances, thereby enabling us to generate more realistic and reliable fantasy point projections.

In developing this method, we employed enhanced for loops to iterate through lists of players, aggregating their fantasy points. This step was essential to calculate the mean fantasy points, laying the groundwork for further calculations. We then used another loop to compute the sum of squared differences from the mean, a critical step in determining the variance of the players' performances. Finally, the standard deviation was derived by taking the square root of this variance.
Reflection

Reflecting on the FRQs and their crossover with our group's final project has been an enlightening experience.

I now am aware of my shortcomings in time management. Piecing bits of code into a cohesive and running program within the 15-minute time limit proved to be difficult. Moreover, I've identified specific technical areas where I need to deepen my understanding. Notably, the concept of interfaces in Java presented a challenge. Grasping the use of interfaces, which are crucial for defining contracts in OOP and ensuring a modular codebase, is something I need to focus on.

I also encountered some cool aspects of programming that were very informative. For instance, learning about the use of StringBuilder was a revelation in terms of efficient string manipulation and memory management. The discovery and application of iterators to safely modify collections while iterating through them was another highlight, as it resolved issues I faced with traditional for loops.

Pointing out the association between the CB problems and the PBL code has helped me realize the immense potential of using our team project as a practical platform for AP exam preparation. Approaching our project with a mindset aligned more closely with the AP Computer Science A exam requirements could aid in improving my coding skills.

As we continue with our project in Trimester 3, I plan to deliberately incorporate features that draw upon elements from all four types of AP CSA exam questions - Methods and Control Structures, Classes, Array/ArrayLists, and 2D Arrays. Especially 2D Arrays, as their traditional forms have not seen use in our project yet.

**Comment by nVarap** [comment]
# Crossover Grading

> Grader: Varaprasad Nibhanupudi

| FRQ | Grade | Comments | 
| ---- | ------ | ----------- |
| 1 | 0.88/0;9 | Doesn't answer the exact questions, but covers more of what is intended close, just really a different class name. For FRQs in the future, remember to answer the actual FRQs |
| 2 | 0.83/0.9 | Demonstrated a poker hand guessing. Similar in idea to what is required (guessing game) but doesn't fully show understanding of the Collegeboard requirements for stringbuilding. Hence, lower score given. |
| 3 | 0.9/0.9 | Did everything as intended, included a little reflection | 
| 4 | 0.9/0.9 | Did everything as intended, included a little reflection | 
| Association | 3.7/4 | Did extra integration, looking for ways to add 2D arrays into their project, used ChatGPT to visualize a possible solution that may be integrated |
